exchanges/populate/src/populateExchange.ts

```diff
@@ -239,117 +245,110 @@ export const addFragmentsToQuery = (
+        warn(
+          'Invalid type: The type ` + type + ` is used with @populate but does not exist.',
+          17
+        );
+      } else {
+        possibleTypes = isAbstractType(type)
+          ? schema.getPossibleTypes(type)
+          : [type];
+      }
 
-            for (let i = 0, l = typeFrags.length; i < l; i++) {
-              const { fragment } = typeFrags[i];
-              const fragmentName = getName(fragment);
-              const usedFragments = getUsedFragments(fragment);
-
-              // Add used fragment for insertion at Document node
-              for (let j = 0, l = usedFragments.length; j < l; j++) {
-                const name = usedFragments[j];
-                if (!existingFragmentsForQuery.has(name)) {
-                  requiredUserFragments[name] = userFragments[name];
-                }
-              }
-
-              // Add fragment for insertion at Document node
-              additionalFragments[fragmentName] = fragment;
-
-              p.push({
-                kind: Kind.FRAGMENT_SPREAD,
-                name: nameNode(fragmentName),
-              });
+      const newSelections = possibleTypes.reduce((p, possibleType) => {
+        const typeFrags = activeTypeFragments[possibleType.name];
+        if (!typeFrags) {
+          return p;
+        }
+
+        for (let i = 0, l = typeFrags.length; i < l; i++) {
+          const { fragment } = typeFrags[i];
+          const fragmentName = getName(fragment);
+          const usedFragments = getUsedFragments(fragment);
+
+          // Add used fragment for insertion at Document node
+          for (let j = 0, l = usedFragments.length; j < l; j++) {
+            const name = usedFragments[j];
+            if (!existingFragmentsForQuery.has(name)) {
+              requiredUserFragments[name] = userFragments[name];
             }
+          }
 
-            return p;
-          }, [] as FragmentSpreadNode[]);
-
-          const existingSelections = getSelectionSet(node);
-
-          const selections =
-            existingSelections.length + newSelections.length !== 0
-              ? [...newSelections, ...existingSelections]
-              : [
-                  {
-                    kind: Kind.FIELD,
-                    name: nameNode('__typename'),
-                  },
-                ];
-
-          return {
-            ...node,
-            directives,
-            selectionSet: {
-              kind: Kind.SELECTION_SET,
-              selections,
-            },
-          };
+          // Add fragment for insertion at Document node
+          additionalFragments[fragmentName] = fragment;
+
+          p.push({
+            kind: Kind.FRAGMENT_SPREAD,
+            name: nameNode(fragmentName),
+          });
+        }
+
+        return p;
+      }, [] as FragmentSpreadNode[]);
+
+      const existingSelections = getSelectionSet(node);
+
+      const selections =
+        existingSelections.length + newSelections.length !== 0
```

Comment:
```suggestion
        existingSelections.length && newSelections.length
```

Reply:
Shouldn't this be `existingSelections.length || newSelections.length` in that case